internal/redispool/ratelimit.go

```diff
@@ -0,0 +1,152 @@
+package redispool
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+var (
+	tokenBucketGlobablPrefix           = "v2:rate_limiters"
+	bucketCapacityConfigKeySuffix      = "bucket_capacity"
+	bucketReplenishmentConfigKeySuffix = "bucket_replenishment_interval_seconds"
+)
+
+type RateLimiter interface {
+	GetTokensFromBucket(ctx context.Context, bucketName, bucketConfigKey string, tokensWanted int) (allowed bool, remianingTokens int, err error)
+	SetTokenBucketReplenishment(ctx context.Context, bucketName, bucketConfigKey string, bucketCapacity, bucketReplenishRateSeconds int) error
+}
+
+type rateLimiter struct {
+	prefix string
+	pool   *redis.Pool
+}
+
+func NewRateLimiter() (RateLimiter, error) {
+	var err error
+
+	pool, ok := Cache.Pool()
+	if !ok {
+		err = errors.New("unable to get redis connection")
+	}
+
+	return &rateLimiter{
+		prefix: tokenBucketGlobablPrefix,
+		pool:   pool,
+	}, err
+}
+
+func (r *rateLimiter) GetTokensFromBucket(ctx context.Context, bucketName, bucketConfigKey string, tokensWanted int) (allowed bool, remianingTokens int, err error) {
+	bucketKey, bucketCapacityKey, bucketReplenishIntervalSecondsKey := r.getRateLimiterKeys(bucketName, bucketConfigKey)
+	result, err := redis.NewScript(3, getTokensFromBucketLuaScript).DoContext(ctx, r.pool.Get(), bucketKey, bucketCapacityKey, bucketReplenishIntervalSecondsKey, tokensWanted)
+	if err != nil {
+		return false, 0, errors.Wrapf(err, "error while getting tokens from bucket %s", bucketName)
+	}
+
+	response, ok := result.([]interface{})
+	if !ok || len(response) != 2 {
+		return false, 0, errors.New("unexpected response from Lua script")
+	}
+
+	allwd, ok := response[0].(int64)
```

Comment:
Nit, rename to `allowed` or `allow`. `allow` is the same amount of characters as `allwd`, but reads much better.
```suggestion
	allow, ok := response[0].(int64)
```